wp-env: Add support for custom WP_HOME port

joemcgill opened the pull request:
## Description
This checks for whether a custom port has been set for `WP_HOME` in the .wp-env.json file before appending the environment's to support local environments where the front end needs to be pointed at another port (e.g., a headless app).

Adds unit test.

## How has this been tested?
Added unit tests. Also testing this config locally and ensured that the "Site Address (URL)" setting in Settings > General reflects the port that was set in my .wp-env.json file.

## Types of changes
New feature (non-breaking change which adds functionality).
Resolves #26481.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

joemcgill committed:
Add support for custom WP_HOME port

This checks for whether a custom port has been set for `WP_HOME` in the .wp-env.json file before appending the environment's to support local environments where the front end needs to be pointed at another port (e.g., a headless app).

Adds unit test.

Resolves #26481

noahtallen commented on packages/env/lib/config/config.js:
```diff
@@ -259,7 +259,12 @@ function withOverrides( config ) {
-				baseUrl.port = config.env[ envKey ].port;
+
+				// Don't overwrite the port of WP_HOME when set.
+				if ( configKey !== 'WP_HOME' || ! baseUrl.port ) {
```
I think we might as well apply this to all of the URLs we check here.

```suggestion
				if ( ! baseUrl.port ) {
```

joemcgill replied:
Thanks @noahtallen. I'd be happy to make this change if you think it's best. The reason that I limited this to `WP_HOME` is that I couldn't think of a legitimate reason why you would want your `WP_SITEURL` or `WP_TESTS_DOMAIN` to point to something other than the port for the WordPress environment. If you prefer limiting this to `WP_HOME` I might rewrite that conditional to be a bit more readable, like:

`if ( ! ( configKey === 'WP_HOME' && baseUrl.port ) ) {`

noahtallen replied:
> I couldn't think of a legitimate reason why you would want your WP_SITEURL or WP_TESTS_DOMAIN to point to something other than the port for the WordPress environment

Ok, that makes sense to me! If you'd like to make it more readable, that sounds good. I would also recommend a quick rebase to see if we can get the e2e tests passing!

noahtallen reviewed (approved):
This looks good to me, just left one comment. I can merge after we figure that out

joemcgill commented:
@noahtallen I've rewritten that conditional and rebased from the latest `master`, but unfortunately tests are already failing in that branch. Doesn't look like any of the failures are related to this PR, so it should be ok.

noahtallen commented:
Yep, confirmed that these e2e failures are the ones that have been plaguing the master branch. Merging now! Thanks for the addition :)
